Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt

```diff
@@ -97,20 +97,33 @@ object CachedKeyFileHolder {
             val deferredQueries: MutableCollection<Deferred<Any>> = mutableListOf()
             keyCache.deleteOutdatedEntries(uuidListFromServer)
             val missingDays = getMissingDaysFromDiff(serverDates)
+            val failedEntryCacheKeys = mutableListOf<String>()
             if (missingDays.isNotEmpty()) {
                 // we have a date difference
                 deferredQueries.addAll(
                     missingDays
                         .map { getURLForDay(it) }
-                        .map { url -> async { url.createDayEntryForUrl() } }
+                        .map { url ->
+                            val cacheKey = url.generateCacheKeyFromString()
+                            try {
+                                async {
+                                    url.createDayEntryForUrl(cacheKey)
+                                }
+                            } catch (e: Exception) {
+                                Timber.v("failed entry: $cacheKey")
+                                failedEntryCacheKeys.add(cacheKey)
+                                throw e
+                            }
+                        }
                 )
             }
             // execute the query plan
             try {
                 deferredQueries.awaitAll()
             } catch (e: Exception) {
+                Timber.v("${failedEntryCacheKeys.size} failed entries ")
                 // For an error we clear the cache to try again
-                keyCache.clear()
+                keyCache.clear(failedEntryCacheKeys)
```

Comment:
`awaitAll` resumes after the first exception is thrown. If we have multiple failed `async` operations, we may already be calling `clear(failedEntryCacheKeys)`  while on another thread `failedEntryCacheKeys.add(cacheKey)` is executed.

So  we wouldn't clear all failed items.
I think we should not rethrow within `async` so that we wait for all operations to finish via `awaitAll`, then check the failedUrls and clear them?